STP module

jbemmel opened the pull request:
* Brings enable/disable STP under Netlab control
* ```stp.protocol```: One of ```stp```(default),```mstp```,```rstp```,```pvrst```, global flavor to run on all supporting nodes
* ```stp.priority``` for setting bridge priority
* ```stp.port_priority``` for selecting between multiple local interfaces

PR modifies FRR vlan template to enable STP before the bridge is enabled, to prevent loops also during startup. A slight violation of modularity, but warranted imho

Cumulus container was found to start sending duplicate BPDUs with both old and new port priority values, requiring a mstpd daemon restart to restore sanity

```
## Platform Support

The following table describes per-platform support of individual STP features:

| Operating system   | STP | MSTP | RSTP | Per-VLAN (R)STP | Enable per port
| ------------------ | :-: | :--: | :--: | :-------------: | :--------------:
| Arista EOS         | ✅  |  ✅  |  ✅  |       ✅        |       ✅      ! Note: STP is enabled by default, using MSTP
| Cumulus Linux      | ✅  |  ❌  |  ✅  |       ❌        |       ✅      ! Note: STP is enabled by default, unless disabled through this module
| FRR                | ✅  |  ❌  |  ❌  |       ✅        |       ❌      ! Note: STP is disabled by default
```

ipspace commented on netsim/modules/stp.py:
```diff
@@ -0,0 +1,71 @@
+  # Check for device support when pvrst is required
+  def node_pre_transform(self, node: Box, topology: Box) -> None:
+
+    if not topology.stp.get('pvrst',False):
```
This will create topology.stp if it doesn't exist. Are you counting on global stp.enable being there?

ipspace commented on netsim/modules/stp.py:
```diff
@@ -0,0 +1,71 @@
+    if not topology.stp.get('pvrst',False):
+      return
+
+    if not 'stp' in node.get('module',[]):
```
I don't think the hook gets called if the node does not use the module

ipspace commented on netsim/modules/stp.py:
```diff
@@ -0,0 +1,71 @@
+    
+    features = devices.get_device_features(node,topology.defaults)
+
+    if not 'stp' in features:
```
This should be detected in the generic node/module compatibility check

ipspace commented on netsim/modules/stp.py:
```diff
@@ -0,0 +1,71 @@
+        'stp')
+      return
+
+    if not features.stp.get('pvrst',False):
```
I would combine this with the first test into a single "can we do it" test. We will need other tests, for example checking **features.stp.mst**.

ipspace commented:
Looks pretty good. The "how do we specify which protocol to use" it the show-stopper, everything else are probably just nits.

jbemmel commented:
> Looks pretty good. The "how do we specify which protocol to use" it the show-stopper, everything else are probably just nits.

Would you prefer a ```stp.protocol``` attribute with the various flavors ```[ "STP", "PVST", "RSTP", "RPVST", "MSTP" ]``` defaulting to "RPVST"?

ipspace commented:
> Would you prefer a `stp.protocol` attribute

Yes

> with the various flavors `[ "STP", "PVST", "RSTP", "RPVST", "MSTP" ]`

Lowercase? We could then use **stp.protocol** to check device features.

> defaulting to "RPVST"?

I try to stay away from STP. Is this a good default?

ipspace commented:
I never claimed to be an STP expert (I tried to stay as far away from it as possible), but MSTP (according to the little I know, and as I understand https://en.wikipedia.org/wiki/Multiple_Spanning_Tree_Protocol) is not per-VLAN STP.

Should that be the case, FRR (implementing VLANs with Linux bridges) cannot run MSTP because each VLAN bridge runs a separate instance of STP and BPDUs are sent tagged with VLAN tags (this https://serverfault.com/questions/824621/linux-bridges-vlans-and-rstp explains it pretty well). It probably does PVST on VLAN interfaces and traditional STP on non-VLAN interfaces.

Also, according to Cumulus Linux 4.4 documentation, it cannot run MSTP but (because we used VLAN-aware bridge) only RSTP: https://docs.nvidia.com/networking-ethernet-software/cumulus-linux-44/Layer-2/Spanning-Tree-and-Rapid-Spanning-Tree/. Also, it effectively tunnels PVST BPDUs.

I would suggest you take a device we can trust to have a decent STP implementation (Arista cEOS, maybe SR Linux) and test interoperability with FRR and CL in VLAN environment. Also, we will not get total interoperability between CL and FRR because our configuration templates used different types of Linux bridges.

jbemmel commented:
I may have been looking at more recent documentation. Removed MSTP from supported protocols, and fixed propagation of ```protocol``` attribute

SRL does not support STP at all, and I don't have easy access to any of the others

jbemmel commented:
* Added EOS support (all flavors)
* Tested interop with FRR on per-VLAN (R)STP - that seems to work fine
* Updated docs to reflect current features per platform

jbemmel commented on netsim/modules/stp.yml:
```diff
@@ -0,0 +1,30 @@
+attributes:
+  global:
+    enable: bool
+    pvrst: bool      # Topology requires Per-VLAN Rapid Spanning Tree (802.1w)
```
Implemented
